simulator/mips/t/unit_test.cpp

```diff
@@ -243,3 +206,454 @@ TEST_CASE( "MIPS32_instr_disasm: Process_Disasm_Branches")
+{
+    MIPS32Instr instr( "mthi");
+    instr.set_v_dst( 0);
+    instr.set_v_src( 0xfee1dead, 0);
+    instr.execute();
+    CHECK( instr.get_v_dst() == 0xfee1dead);
+}
+
+// Not implemented in SPIM TT
+TEST_CASE( "MIPS32_instr: mflo")
+{
+    MIPS32Instr instr( "mflo");
+    instr.set_v_dst( 0);
+    instr.set_v_src( 0xfee1dead, 0);
+    instr.execute();
+    CHECK( instr.get_v_dst() == 0xfee1dead);
+}
+
+// Not implemented in SPIM TT
+TEST_CASE( "MIPS32_instr: mtlo")
+{
+    MIPS32Instr instr( "mtlo");
+    instr.set_v_dst( 0);
+    instr.set_v_src( 0xfee1dead, 0);
+    instr.execute();
+    CHECK( instr.get_v_dst() == 0xfee1dead);
+}
+
+// Not implemented in SPIM TT
+TEST_CASE( "MIPS64_instr: dsllv")
+{
+    MIPS64Instr instr( "dsllv");
+
+    instr.set_v_src( 1, 0);
+    instr.set_v_src( 0, 1);
+    instr.execute();
+    CHECK( instr.get_v_dst() == 1);
+
+    instr.set_v_src( 1, 0);
+    instr.set_v_src( 1, 1);
+    instr.execute();
+    CHECK( instr.get_v_dst() == 2);
+
+    instr.set_v_src( 1, 0);
+    instr.set_v_src( 32, 1);
+    instr.execute();
+    CHECK( instr.get_v_dst() == 0x100000000);
+
+    instr.set_v_src( 1, 0);
+    instr.set_v_src( 64, 1);
+    instr.execute();
+    CHECK( instr.get_v_dst() == 1);
+}
+
+// Not implemented in SPIM TT
+TEST_CASE( "MIPS64_instr: dsrlv")
+{
+    MIPS64Instr instr( "dsrlv");
+
+    instr.set_v_src( 1, 0);
+    instr.set_v_src( 0, 1);
+    instr.execute();
+    CHECK( instr.get_v_dst() == 1);
+
+    instr.set_v_src( 1, 0);
+    instr.set_v_src( 1, 1);
+    instr.execute();
+    CHECK( instr.get_v_dst() == 0);
+
+    instr.set_v_src( 0x10000000, 0);
+    instr.set_v_src( 4, 1);
+    instr.execute();
+    CHECK( instr.get_v_dst() == 0x01000000);
+
+    instr.set_v_src( 0x8000000000000000, 0);
+    instr.set_v_src( 4, 1);
+    instr.execute();
+    CHECK( instr.get_v_dst() == 0x0800000000000000);
+}
+
+// Not implemented in SPIM TT
+TEST_CASE( "MIPS64_instr: dsrav")
+{
+    MIPS64Instr instr( "dsrav");
+
+    instr.set_v_src( 1, 0);
+    instr.set_v_src( 0, 1);
+    instr.execute();
+    CHECK( instr.get_v_dst() == 1);
+
+    instr.set_v_src( 1, 0);
+    instr.set_v_src( 1, 1);
+    instr.execute();
+    CHECK( instr.get_v_dst() == 0);
+
+    instr.set_v_src( 0x100000000000, 0);
+    instr.set_v_src( 4, 1);
+    instr.execute();
+    CHECK( instr.get_v_dst() == 0x010000000000);
+
+    // Little endian
+    instr.set_v_src( 0xff00000000000000, 0);
+    instr.set_v_src( 4, 1);
+    instr.execute();
+    CHECK( instr.get_v_dst() == 0xfff0000000000000);
+}
+
+// ******************** I and J INSTRUCTIONS ***************************
+// 1 done
+
+TEST_CASE( "MIPS32_instr: beq")
+{
+    MIPS32Instr instr( "beq");
+
+    instr.set_new_PC( instr.get_PC() + 4);
+    instr.set_v_src( 0, 0);
+    instr.set_v_src( 0, 1);
+    instr.set_v_imm( 1);
+    instr.execute();
+    CHECK( instr.get_new_PC() == instr.get_PC() + 8);
+
+    instr.set_new_PC( instr.get_PC() + 4);
+    instr.set_v_src( -1, 0);
```

Comment:
You cannot just write "-1" to uint32.
Instead, define a constant in the outer scope:
```c++
static constexpr const uint32 MINUS_ONE32 = all_ones<uint32>();
```

Reply:
Yes you can, but IMO this is less readable (how many 'f's are there?)